tests/draft2019-09/pattern.json

```diff
@@ -49,11 +49,27 @@
         "description": "pattern is not anchored",
         "schema": {"pattern": "a+"},
         "tests": [
+            {
+                "description": "doesn't match a substring",
+                "data": "xxyy",
+                "valid": false
+            },
             {
                 "description": "matches a substring",
                 "data": "xxaayy",
                 "valid": true
             }
         ]
+    },
+    {
+        "description": "pattern is malformed",
+        "schema": {"pattern": "hel{1,",
+        "tests": [
+            {
+                "description": "invalid regex does not match",
+                "data": "hello",
+                "valid": false
```

Comment:
I'm not expecting the metaschema to use a format validation.  Even if no validation is done on the schema at all, it should still return an invalid result at runtime.

Reply:
That may not be the case. An implementation is free to blow up entirely in this case if you give it an invalid schema and it isn't validating it.

Reply:
Isn't "my validator blows up" covered by `"valid":false`?  This is a boolean, after all! :)

Reply:
Note: this isn't a rhetorical question. How *do* implementations handle "oops I blew up" when evaluating the test suite?

I'll answer for my case -- in the loop that iterates over all test cases, if an exception was thrown, the test runner catches it and generates an invalid result.

Reply:
[This answer is more for posterity for people looking at this PR.] This is one of the reasons that it is preferred for schemas in the tests to be valid. Then, there's no ambiguity in how test runners run these tests. This PR demonstrates why we need meta-schema tests and why they need to be handled differently than regular tests.

Reply:
How about a separate optional/ directory for test cases that fall into the "the implementation's behaviour is undefined" or "the schema itself was invalid, so the implementation might blow up" category?  that way the main directory can stick to strictly testing schemas that are conformant to the draftX metaschema.

Reply:
> Some test runners just don't run the whole group of tests if the schema fails...

That's quite reasonable, since the schema is constant across that entire group of tests, so the same explosion is going to happen equally for all of them.

Reply:
> How about a separate optional/ directory for test cases that fall into the "the implementation's behaviour is undefined" or "the schema itself was invalid, so the implementation might blow up" category? that way the main directory can stick to strictly testing schemas that are conformant to the draftX metaschema.

That's exactly what we already do yeah, sorry, that was what I was suggesting earlier with "this should go in optional" -- you're right, it'd be even better if there was *2* such locations (or sometimes I consider 3) -- but one location for "you may optionally implement this behavior" and one for "you may be in a certain situation, in which case you *should* implement this behavior, but the test suite doesn't know whether you are in or not in that situation so we call it optional but you need to know and apply it accordingly"

But for now everything in both of those categories is in one place -- optional.

(Lemme know if that makes sense, this is a tricky point -- but yes `valid: false` is today definitely not the same as "may cause your CI machine to segfault or blow up when you run the tests")

Reply:
how about `optional/undefined-behaviour/` for these sorts of tests?  (or I guess `optional/undefined-behavior/` to satisfy the yanks..) :)

Reply:
> That's quite reasonable, since the schema is constant across that entire group of tests, so the same explosion is going to happen equally for all of them.
That's not reasonable if some of the tests happen to be marked as "valid=true".

Reply:
If the schema is invalid, how could any of the tests be valid=true?

Reply:
There seems to be a confusion between writing the tests and implementing a test runner. That case, while not desired, is possible.

Reply:
The whole point here is that the test suite is not designed to run tests having invalid schemas in `schema`.

Reply:
Show me a test case with an invalid schema where "valid":true is the expected result and I'll retract my statement :)

FWIW, I already maintain a test runner, that I adopted from @Relequestual. I'm not trying to write another one :)

Reply:
The issue is that implementation vs. tests is still being confused. Invalid schema does _not_ mean “valid=false”. It means that a test or tests are not even executed. You cannot make the assumption that “valid=false”.

Reply:
There’s a difference between “shouldn’t exist“ and “don’t have to account for all the cases”.

Reply:
As well, assuming that “valid=false” for failed schemas changes the meaning of “valid” for the tests in that suite. It’s also bad practice to have meanings of fields change depending on other variables.

Reply:
Perhaps there should be another flag to indicate invalid schemas (apart from `valid: false`), perhaps in a separate directory/subdir, where the impl is free to blow up on schema compilation (like it should imo)?

That would be different from `valid: false`, as the impl can either follow this and return false, because none strings match invalid pattern, or it can as well reject the schema alltogehter which should also be a valid behavior.

Reply:
Yes, we've been discussing this across a few different PRs and in slack.  There hasn't been any consensus yet as to how to declare this in tests (with an extra flag? a separate directory under optional/? both of these?) or how to document how implementations should deal with it.

Reply:
I want to merge the outstanding PRs on tests in the "normal" category before addressing this. I.e. #354  and the PRs being split off of it. Let's get those merged, then we can hit cases that may require a structural change. Could definitely use more eyes reviewing those PRs so we can get them off the queue.